java/code/src/com/redhat/rhn/common/hibernate/HibernateFactory.java

```diff
@@ -794,36 +817,47 @@ protected static void executeCallableMode(String name, String mode, Map params)
     @SuppressWarnings("unchecked")
     protected static <T, ID> List<T> findByIds(List<ID> ids, String queryName,
             String idsParameterName, Map<String, Object> parameters) {
-        Session session = HibernateFactory.getSession();
-        org.hibernate.query.Query<T> query = session.getNamedQuery(queryName);
+        Query<T> query = HibernateFactory.getSession().getNamedQuery(queryName);
+
         parameters.entrySet().stream().forEach(entry -> {
             query.setParameter(entry.getKey(), entry.getValue());
         });
-        List<T> results = new LinkedList<T>();
 
-        if (ids.size() == 0) {
-            return results;
-        }
+        BinaryOperator<List<T>> binaryOperator = new BinaryOperator<List<T>>() {
+            @Override
+            public List<T> apply(List<T> a, List<T> b) {
+                a.addAll(b);
+                return a;
+            }
+        };
 
-        if (ids.size() < LIST_BATCH_MAX_SIZE) {
-            query.setParameterList(idsParameterName, ids);
-            return query.getResultList();
-        }
+        return splitAndExecuteQuery(ids, idsParameterName, query, query::getResultList,
+                new ArrayList<T>(), binaryOperator);
+    }
 
-        List<ID> blockOfIds = new LinkedList<ID>();
-        for (ID sid : ids) {
-            blockOfIds.add(sid);
-            if (blockOfIds.size() == LIST_BATCH_MAX_SIZE - 1) {
-                query.setParameterList(idsParameterName, blockOfIds);
-                results.addAll(query.getResultList());
-                blockOfIds = new LinkedList<ID>();
-            }
-        }
-        // Deal with the remainder:
-        if (blockOfIds.size() > 0) {
-            query.setParameterList(idsParameterName, blockOfIds);
-            results.addAll(query.getResultList());
+    /**
+     * Splits a list of elements in batches of LIST_BATCH_MAX_SIZE and execute a query for each batch.
+     * @param <T> the return type
+     * @param <E> the type of the elements in the list parameter
+     * @param <R> the type of the returned objects by the query
+     * @param list the list of parameters to search for
+     * @param parameterName the name of the parameter to match the parameters in the list
+     * @param query the query to be executed
+     * @param queryFunction the function to be call on the query
+     * @param accumulatorInitValue the initial value for the result accumulator
+     * @param accumulatorOp the operation for the result accumulator
+     * @return an accumulated result of executing the query
+     */
+    protected static <E, T, R> T splitAndExecuteQuery(List<E> list, String parameterName,
+            Query<R> query, Supplier<T> queryFunction, T accumulatorInitValue, BinaryOperator<T> accumulatorOp) {
+        int size = list.size();
+
+        List<List<E>> batches = new ArrayList<List<E>>();
+        for (int i = 0; i < size; i += LIST_BATCH_MAX_SIZE) {
```

Comment:
It would be cool to rewrite this using `Stream`s exclusively, without relying on this `for` loop.

More of a functional gymnastics exercise than anything of incredibly high practical use, still I'd recommend giving that a try :wink: